File: packages/fsm-compiler-ts/src/checkers/check_fn.py
#!/usr/bin/env python3
"""Check that a named function is defined in a Python source file.

Usage: python3 check_fn.py <filepath> <fn_name>
Exit 0 if the function is found, 1 if not, 2 on bad arguments or parse error.
"""

import ast
import sys


def _has_fn(source: str, fn_name: str) -> bool:
    for node in ast.walk(ast.parse(source)):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            if node.name == fn_name:
                return True
    return False


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: check_fn.py <filepath> <fn_name>", file=sys.stderr)
        sys.exit(2)

    filepath, fn_name = sys.argv[1], sys.argv[2]
    try:
        with open(filepath) as f:
            source = f.read()
    except OSError as e:
        print(f"Cannot read {filepath}: {e}", file=sys.stderr)
        sys.exit(2)

    try:
        found = _has_fn(source, fn_name)
    except SyntaxError as e:
        print(f"Syntax error in {filepath}: {e}", file=sys.stderr)
        sys.exit(2)

    if found:
        sys.exit(0)
    else:
        print(f"Function '{fn_name}' not found in {filepath}", file=sys.stderr)
        sys.exit(1)

File: packages/fsm-compiler-ts/src/checkers/check_fn.rs
// Check that a named public function is declared in a Rust source file.
//
// Compile once: rustc check_fn.rs -o check_fn_rust
// Usage:        ./check_fn_rust <filepath> <fn_name>
// Exit 0 if found, 1 if not, 2 on bad arguments or I/O error.
use std::{env, fs, process};

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: check_fn_rust <filepath> <fn_name>");
        process::exit(2);
    }
    let filepath = &args[1];
    let fn_name = &args[2];

    let content = match fs::read_to_string(filepath) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Cannot read {}: {}", filepath, e);
            process::exit(2);
        }
    };

    // Match `pub fn fn_name(` and `pub async fn fn_name(`
    let patterns = [
        format!("pub fn {}(", fn_name),
        format!("pub fn {} (", fn_name),
        format!("pub async fn {}(", fn_name),
        format!("pub async fn {} (", fn_name),
    ];
    for pat in &patterns {
        if content.contains(pat.as_str()) {
            process::exit(0);
        }
    }

    eprintln!("Function '{}' not found in {}", fn_name, filepath);
    process::exit(1);
}
